### PR TITLE
feat: add webFrameMain.send() / webFrameMain.postMessage()

### PR DESCRIPTION
**Pull request by miniak**
#### Description of Change
Add `webFrameMain.send()` / `webFrameMain.postMessage()`.

#### Checklist
- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes
Notes: Added `webFrameMain.send()` / `webFrameMain.postMessage()`.

### CODE REVIEW

**Comment by canibaltrol** [comment]
Can u stop, send spam 24/7 tnx.

2020 m. gruod. 3 d., kt 04:03, Milan Burda <notifications@github.com> rašė:

File: docs/api/web-frame-main.md [hunk 1]
```diff
@@ -89,6 +89,20 @@ this limitation.
 
 Returns `boolean` - Whether the reload was initiated successfully. Only results in `false` when the frame has no history.
 
+#### `frame.send(channel, ...args)`
```
Comment (nornagon) [inline docs/api/web-frame-main.md:92]: We should have `frame.postMessage` as well, so MessagePorts can be passed.

File: lib/browser/init.ts [hunk 2]
```diff
@@ -142,6 +142,9 @@ require('@electron/internal/browser/api/protocol');
 // Load web-contents module to ensure it is populated on app ready
 require('@electron/internal/browser/api/web-contents');
 
+// Load web-frame-main module to ensure it is populated on app ready
+require('@electron/internal/browser/api/web-frame-main');
```
[the remark this answers is not in the document]
Comment (miniak) [reply lib/browser/init.ts:146]: we need to modify the prototype of WebFrameMain to add the `send()` method, which is calling the internal `_send()`

File: docs/api/web-frame-main.md (hunk 1 above)
Comment (miniak) [reply docs/api/web-frame-main.md:92]: should I do it here or in a separate PR? I would merge this as it's ready a do a follow-up

File: shell/browser/api/electron_api_web_frame_main.cc [hunk 3]
```diff
@@ -114,6 +119,63 @@ bool WebFrameMain::Reload(v8::Isolate* isolate) {
...
+
+  if (!CheckRenderFrame())
+    return;
+
+  mojo::AssociatedRemote<mojom::ElectronRenderer> electron_renderer;
+  render_frame_->GetRemoteAssociatedInterfaces()->GetInterface(
+      &electron_renderer);
+  electron_renderer->ReceivePostMessage(channel,
+                                        std::move(transferable_message));
+}
```
Comment (nornagon) [inline shell/browser/api/electron_api_web_frame_main.cc:177]: Can we share the logic of these two functions with the WebContents version somehow?

File: lib/browser/api/web-contents.ts [hunk 4]
```diff
@@ -134,37 +138,30 @@ WebContents.prototype.send = function (channel, ...args) {
...
-  } else if (!(typeof frame === 'number' || Array.isArray(frame))) {
-    throw new Error('Missing required frame argument (must be number or array)');
+
+function getWebFrame (contents: Electron.WebContents, frame: number | [number, number]) {
+  if (typeof frame === 'number') {
+    return webFrameMain.fromId(contents.mainFrame.processId, frame);
+  } else if (Array.isArray(frame) && frame.length === 2 && frame.every(value => typeof value === 'number')) {
+    return webFrameMain.fromId(frame[0], frame[1]);
+  } else {
+    throw new Error('Missing required frame argument (must be number or [processId, frameId])');
```
Comment (nornagon) [inline lib/browser/api/web-contents.ts:155]: We should guard against a `null` return value from `webFrameMain.fromId` here.

**Review by nornagon — APPROVED** [review]

File: lib/browser/api/web-contents.ts (hunk 4 above)
Comment (miniak) [reply lib/browser/api/web-contents.ts:155]: I will do that, but need this #27068 to land first

Comment (miniak) [reply lib/browser/api/web-contents.ts:155]: `WebContents.prototype.sendToFrame` and `WebContents.prototype._sendToFrameInternal` now checks the return value of `getWebFrame`

**Review by zcbenz — APPROVED** [review]

File: docs/api/web-frame-main.md [hunk 5]
```diff
@@ -101,6 +101,47 @@ Works like `executeJavaScript` but evaluates `scripts` in an isolated context.
...
+// Main process
+const { port1, port2 } = new MessageChannelMain()
+webContents.mainFrame.postMessage('port', { message: 'hello' }, [port1])
+
+// Renderer process
+ipcRenderer.on('port', (e, msg) => {
+  const [port] = e.ports
+  // ...
+})
+```
```
Comment (samuelmaddock) [inline docs/api/web-frame-main.md:143]: ❤️ this

**Review by samuelmaddock — APPROVED** [review]
👍

**Comment by jeremypointer7** [comment]
Is that what I need to do?



Sent from my Metro By T-Mobile 4G LTE Android Device



-------- Original message --------

**Review by samuelmaddock — APPROVED** [review]
API LGTM

**Review by jkleinsc — APPROVED** [review]

**Pull request merged**